api/users.go

```diff
@@ -53,35 +60,383 @@ func (self *ApiServer) SetPassword(
 	return &emptypb.Empty{}, users_manager.SetUser(ctx, user_record)
 }
 
+type userByName []*api_proto.VelociraptorUser
+
+func (s userByName) Len() int {
+    return len(s)
+}
+func (s userByName) Swap(i, j int) {
+    s[i], s[j] = s[j], s[i]
+}
+func (s userByName) Less(i, j int) bool {
+    return s[i].Name < s[j].Name
+}
+
 func (self *ApiServer) GetUsers(
 	ctx context.Context,
-	in *emptypb.Empty) (*api_proto.Users, error) {
+	in *api_proto.GetUsersRequest) (*api_proto.Users, error) {
 
 	users_manager := services.GetUserManager()
 	user_record, org_config_obj, err := users_manager.GetUserFromContext(ctx)
 	if err != nil {
 		return nil, Status(self.verbose, err)
 	}
 
-	permissions := acls.READ_RESULTS
-	perm, err := acls.CheckAccess(org_config_obj, user_record.Name, permissions)
+	perm, err := acls.CheckAccess(org_config_obj, user_record.Name, acls.SERVER_ADMIN)
 	if !perm || err != nil {
 		return nil, status.Error(codes.PermissionDenied,
 			"User is not allowed to enumerate users.")
 	}
 
+	filterOrgs := true
+
+	if isMemberOfRootOrg(user_record.Orgs) {
+		filterOrgs = false
+	} else {
+		ok, err := acls.CheckAccess(org_config_obj, user_record.Name, acls.ORG_ADMIN)
+		if err != nil {
+			return nil, err
+		}
+		if ok {
+			filterOrgs = false
+		}
+	}
+
 	result := &api_proto.Users{}
 
 	users, err := users_manager.ListUsers(ctx)
 	if err != nil {
 		return nil, Status(self.verbose, err)
 	}
 
+	if filterOrgs {
+		newUsers := []*api_proto.VelociraptorUser{}
+		callerOrgs := map[string]bool{}
+		for _, org := range user_record.Orgs {
+			callerOrgs[org.Id] = true
+
+		}
+
+		for _, user := range users {
+			for _, org := range user.Orgs {
+				_, found := callerOrgs[org.Id]
+				if found {
+					newUsers = append(newUsers, user)
+					break
+				}
+			}
+		}
+		users = newUsers
+	}
+
+	if in.WithRoles {
+		for _, user := range users {
+			user.Permissions, err = acls.GetPolicy(org_config_obj, user.Name)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	sort.Sort(userByName(users))
```

Comment:
You can use sort.Slice() a bit easier here so you dont need another type.

Reply:
Ack. Integrated.